File: javascript/lesson 6/inOrder.js
// https://leetcode.com/problems/binary-tree-inorder-traversal/

function inorderTraversal(root) {
    const result = [];
    inOrder(root, result);
    return result;
}

function inOrder(node, result) {
    if (!node) return;

    inOrder(node.left, result);
    result.push(node.val);
    inOrder(node.right, result);
}

File: kotlin/lesson 6/InOrder.kt
// https://leetcode.com/problems/binary-tree-inorder-traversal/

class Solution {
    fun inorderTraversal(root: TreeNode?): List<Int> {
        val result = mutableListOf<Int>()
        inOrder(root, result)
        return result
    }
    
    fun inOrder(node: TreeNode?, result: MutableList<Int>) {
        if (node == null) return

        inOrder(node.left, result)
        result.add(node.`val`)
        inOrder(node.right, result)
    }
}
